### DISCUSSION TITLE
Delete multiple records event hook

### DISCUSSION THREAD

**Discussion by adibfarrasy** [pocketbase/pocketbase#5584] [Q&A] [answered]
I have 3 collections, named `products`, `product_units`, and `cart_items`, with:
- `products` record having multiple `product_units` relations, cascade delete = true
- `cart_items` record having single `product_units` relation, cascade delete = true

the desired behaviors are:
- if a `products`  record is deleted, the multiple relations `product_units` records and the associated `cart_items` records get deleted too. 
- if `product_units` records are deleted, the `products` records containing all of the relations and `cart_items` records are deleted.

This was okay if the `product_units` records are deleted (2nd case), but how do I trigger the chain deletion on deleting `products` record?

I tried to do 
```
        app.OnModelBeforeDelete("products").Add(func(e *core.ModelEvent) error {
		product, err := app.Dao().FindRecordById("products", e.Model.GetId())
		if err != nil {
			return err
		}

		app.Dao().ExpandRecord(product, []string{"product_units"}, nil)

		productUnitIDs := product.GetStringSlice("product_units")
		puidFilter := lo.Map(productUnitIDs, func(id string, _ int) string {
			return fmt.Sprintf("product_unit.id = '%s'", id)
		}) // samber/lo library, just a fancy for-loop to map slices

		cartItems, err := app.Dao().FindRecordsByFilter(
			"cart_items",
			strings.Join(puidFilter, " || "),
			"-created",
			len(puidFilter),
			0,
		)
		if err != nil {
			return err
		}

		for _, item := range cartItems {
			if err := app.Dao().DeleteRecord(item); err != nil {
				return err
			}
		}

		productUnits := product.ExpandedAll("product_units")
		for _, unit := range productUnits {
			if err := app.Dao().DeleteRecord(unit); err != nil {
				return err
			}
		}

		return nil
	})
```
but it seems to deadlock when I deleted the `products` record. I tried experimenting with the code and cascade delete = false too, to no avail. any suggestion?

**Comment by ganigeorgiev** [comment, marked as answer]
I'm having difficulties understanding what is the end goal here, but the deadlock is happening because the delete operation is executed in a transaction and when using the DB hooks (aka. `OnModel*`) you should use the event Dao (`e.Dao`) instead of the `app.Dao()` because we allow only 1 writer/transaction at a time (_note that the `Dao` abstraction is removed in the upcoming v0.23.0 release_).

**Reply by ganigeorgiev** [reply]
_As a side-note: if I understand your description correctly and if a single `product_units` can be associated to only 1 product, then you can delete the multiple product relation field and add a single products `relation` in the `products_units` collection and this should take care for automatically deleting `products_units` in case of `products` deletion (if that is the desired behavior)._

**Comment by adibfarrasy** [comment]
ok I missed that the event object passed actually have its own Dao. It works as expected now, without having to update the collection schema. Thanks @ganigeorgiev :+1:
